### PR TITLE
Upsert

### LINKED ISSUE #298: Add upsert ability to SaveOperations
Sometimes you want to save a record, sometimes you want to create a record. But there's also some cases when you're not sure which one you need, so an `upsert`. 

```crystal
SaveUser.upsert do |op, user|
  #...
end
```

The SQL would end up looking something like this using `ON CONFLICT`
```
INSERT INTO users (id, name) VALUES (1, 'Billy') ON CONFLICT DO UPDATE SET name='Billy';
```

### PR DESCRIPTION
**Pull request by paulcsmith**
Closes #298

Needs

- [x] Document the macro `upsert_lookup_columns`
- [x] Document how to use `upsert` and `upsert!`
- [x] Create in luckyframework/website to add section in the guides https://github.com/luckyframework/website/issues/716
- [x] Extract a #save method that accepts a block so we don't have duplicate code

### CODE REVIEW

**Comment by wontruefree** [comment]
So this does not use on conflict.  Is that going to be added in a later PR?

**Comment by paulcsmith** [comment]
@wontruefree Yep! Ideally it will use `on conflict`, but that'll require adding new methods to the query builder so figured we can at least get this out there in the meantime, then improve it later

File: src/avram/save_operation.cr
Comment (wontruefree) [inline src/avram/save_operation.cr]: Why is this exposed?

Comment (paulcsmith) [reply src/avram/save_operation.cr]: So that it can be set in the upsert if it finds an existing record

Comment (wontruefree) [reply src/avram/save_operation.cr]: I see it now.

**Comment by confact** [comment]
@paulcsmith A idea i have for find_or_create is to maybe add posibility to use a block.
then it use the parameters to find it and if not find call the block to add those fields. It makes more sense in my head using that.

as example:
```
email = "hakan@example.com"
User.find_or_create(email: email) do |u|
  u.email = email
  u.username = "confact"
  u.role = :user 
  u.name = params["name"]
end
```

**Comment by paulcsmith** [comment]
@confact I like that syntax, but for a couple of reasons it is tricky to get type-safe and to work with param parsing. But I like the idea. I tried a few APIs before and couldn't get something that worked, but I'll think on it and see if we can do something closer to that style.

**Comment by stephendolan** [comment]
@jwoertink I've merged in `master` to this branch and resolved the conflicts/updates necessary. What are the next steps to get this merge-able?

I'm happy to take some guidance and attempt to tackle finishing this one up myself, too!

File: src/avram/save_operation.cr
Comment (jwoertink) [inline src/avram/save_operation.cr:43]: This is duplicated.

File: src/avram/save_operation.cr
Comment (jwoertink) [inline src/avram/save_operation.cr:42]: Let's move the `getter :record` to this so we're not defining that getter twice.

File: src/avram/unique_columns.cr
Comment (jwoertink) [inline src/avram/unique_columns.cr:14]:
```suggestion
    def self.upsert!(*args, **named_args) : T
```

File: src/avram/unique_columns.cr
Comment (jwoertink) [inline src/avram/unique_columns.cr:3]:
```suggestion
    def self.find_or_create!(*args, **named_args) : T
```

**Review by jwoertink — CHANGES_REQUESTED** [review]
@stephendolan Overall this looks good. I left a few suggestions to fix. The only major thing is we don't have the block forms of these methods. So we can either add those in, or roll with this as is and add the block forms as a separate PR. I'm good either way.

**Comment by stephendolan** [comment]
@jwoertink I opted not to introduce all the block variants in this PR so that we can get it cleaned out of the queue and at least have the basics available for folks in the next release.

If we merge this as-is, I can follow up with an issue to document the need for those additional overloads.

File: spec/operations/save_operation_spec.cr
Comment (matthewmcgarvey) [inline spec/operations/save_operation_spec.cr:179]: This took a while to understand. We're saying find_or_create but it's not using all the values we pass to the method

**Review by matthewmcgarvey — COMMENTED** [review]
I feel iffy about this. Here's some things going through my mind, they might or might not be worth anything, just what I'm thinking:
- As I commented, `unique_columns` is weird to me because the call site provides data that isn't used in the fetch. It feels disjointed
- It's introducing two methods/macros that have to be used together
- We are depending on the query object that we generated

Because we separate out queries and operations I'm hesitant to believe this is the best way to go so I looked into Elixir's Ecto ORM. They don't have a `find_or_create` method or an `upsert`. Instead, they provide options on their `Repo.insert` function to handle conflicts.

https://hexdocs.pm/ecto/Ecto.Repo.html#c:insert/2

Can we talk before merging this?

**Comment by jwoertink** [comment]
Yeah, I'm ok with holding off on this.

**Comment by stephendolan** [comment]
I'm definitely cool with a different approach, as long as this doesn't sit for another year untouched 😆 .

Having just implemented this manually in two apps, it definitely feels like an area where it'll be easy for end-users of Lucky to make mistakes until we provide the capability.

**Comment by confact** [comment]
Yea, I think it is good to have something like this. I am doing find and check if nil to create now in many places so it would for sure save a lot of code.

But the current flow is confusing to me. I am not sure what it is using to find it and what it will update. is it some of them or all of them?

If we find a good way to handle that in a good way it would be good. But I understand the issue @paulcsmith talked about with the parameters and block, so a ruby on rails-like method might be hard.

Not sure exactly how to fix it. I think it is a good call to wait out and think it through.

File: src/avram/save_operation.cr
Comment (wontruefree) [inline src/avram/save_operation.cr:38]: hmmm.  I see why you did this but should the be a part of the public api?

File: spec/operations/save_operation_spec.cr
Comment (jwoertink) [inline spec/operations/save_operation_spec.cr:61]: I like how this reads.

File: spec/operations/save_operation_spec.cr
Comment (jwoertink) [inline spec/operations/save_operation_spec.cr:242]: So this says "Look up a user by name = "Rich" and nickname = "R." which doesn't exist, right? With the bang method, should this raise an exception like `Avram::RecordNotFound`? Or maybe `user` should be `nil` here?

File: src/avram/save_operation.cr
Comment (jwoertink) [reply src/avram/save_operation.cr:38]: @wontruefree you mean maybe it should be like `public getter` and `private setter`? I could see something like that

File: src/avram/upsert.cr
Comment (jwoertink) [inline src/avram/upsert.cr:8]: @wontruefree Ah, I think it's because here it would need to be public somehow. I guess an alternative could be doing something like `operation.set_record(_record: existing_record)`, but that probably just muddles up the code more than we save by just making it public and not documenting it. What do you think?

File: src/avram/upsert.cr
Comment (jwoertink) [inline src/avram/upsert.cr:2]: Out of curiosity, what does the error look like if you misspell a column name, or add a name here that doesn't exist? `upset_lookup_columns :bloopity`

**Review by jwoertink — COMMENTED** [review]
This is looking nice! Just a few questions, but I dig the direction.

File: src/avram/save_operation.cr
Comment (paulcsmith) [reply src/avram/save_operation.cr:38]: Good point. _probably_ not, or at the very least should not be documented

File: spec/operations/save_operation_spec.cr
Comment (paulcsmith) [reply spec/operations/save_operation_spec.cr:242]: With the bang version it should only raise if the record is invalid. Otherwise it will create a new one if the record is not found. I should add a test for that though!

File: src/avram/upsert.cr
Comment (paulcsmith) [reply src/avram/upsert.cr:8]: I'll keep it public but not document. There may be other extensions/shards that need to hook into it. Having it there, but not documented should leave it open, but not leave regular users confused about what it is for

Comment (jwoertink) [reply src/avram/upsert.cr:8]: That sounds good.

File: src/avram/upsert.cr
Comment (paulcsmith) [reply src/avram/upsert.cr:2]: Pretty good actually! Shows the macro that called it and gives a nice suggestion (assuming it is a typo and not totally wrong)

Screen Shot 2021-07-12 at 2 10 33 PM

Comment (jwoertink) [reply src/avram/upsert.cr:2]: boom! That's way nice 👍

File: src/avram/needy_initializer_and_save_methods.cr
Comment (paulcsmith) [inline src/avram/needy_initializer_and_save_methods.cr:171]: This was moved to `SaveOperation#save` so we don't need to remember to call it in multiple places

File: src/avram/upsert.cr
Comment (jwoertink) [inline src/avram/upsert.cr:22]:
```suggestion
  # class User < BaseModel
```

File: src/avram/upsert.cr
Comment (jwoertink) [inline src/avram/upsert.cr:100]: Just thinking like a newbie here, this might lead me to believe I need to do something like:

```crystal
SaveUser.upsert_lookup_columns
SaveUser.upsert!(...)
```

Maybe we can word it like "Please include the `upsert_lookup_columns` macro in #{@type} before using '{{ methods.id }}'" ?

**Review by jwoertink — APPROVED** [review]
This is great stuff! Overall, the implementation seems fairly simple which is nice. Just a few small wording suggestions, but if you're cool with it, merge away!

File: src/avram/upsert.cr
Comment (paulcsmith) [reply src/avram/upsert.cr:100]: Good idea! THat makes a lot of sense

**Review by matthewmcgarvey — APPROVED** [review]

**Pull request merged**
